### PR TITLE
Add ability to prevent exception propagation if building as part of ORT when ORT has exceptions disabled

### PR DESCRIPTION
**Pull request by skottmckay**
Add ability to prevent exception propagation with top level try/catch hander macros.

If combined build with ORT has exceptions disabled in ORT but ort-ext has an operator that requires exceptions (e.g. SentencePiece), we enable exceptions in ort-ext but prevent them propagating up via try/catch in the entry points that ORT can call. These are:
  - RegisterCustomOps
  - CustomOpBase constructor 
  - CustomOpBase Compute

Removed some places in CustomOpApi that threw if OpKernelInfo* was nullptr but standardizing all kernels to store the OpKernelInfo provided in the ctor. Having one standard CreateKernel meant a lot of overrides could be removed.

Added unit tests
  - need to validate on more platforms and add CI for build where we don't want to allow exceptions to propagate

### CODE REVIEW

File: operators/cv2/imgcodecs/imdecode.hpp
Comment (skottmckay) [inline operators/cv2/imgcodecs/imdecode.hpp:36]: Suggest hiding whitespace diffs as clang-format has run when saving the edits causing lots of diffs. 
Other whitespace diffs are from conforming to 120 char limit but those might not get hidden.

image

File: CMakeLists.txt
Comment (skottmckay) [inline CMakeLists.txt:193]: OCOS_PREVENT_EXCEPTION_PROPAGATION may be clearer

File: pyop/pykernel.h
Comment (wenbingl) [inline pyop/pykernel.h:64]: there are some extra arguments here (opdef_->obj_id...), doesn't this removal cause some issues?

Comment (skottmckay) [reply pyop/pykernel.h:64]: Good catch. Fixed.

File: CMakeLists.txt
Comment (skottmckay) [reply CMakeLists.txt:185]: I copied from the ORT code [here](https://github.com/microsoft/onnxruntime/blob/abb51ec97599adee62c6c90f950a719ee778b974/cmake/adjust_global_compile_flags.cmake#L161)

Will try REPLACE as that seems simpler.

File: CMakeLists.txt
Comment (skottmckay) [reply CMakeLists.txt:167]: I believe the leading '_' is an internal value not expected to be set directly by the user. i.e. isn't in the list of public options specified around line 43.

Do you have an alternative naming suggestion? I was defaulting to OFF mainly so the name of the value set is visible here and letting SetExceptionsRequired update to the correct value. AFAIK cmake doesn't allow returning a value from a function so something like `set(_OCOS_EXCEPTIONS_REQUIRED SetExceptionsRequired())` isn't possible. That said, noexceps_ops just went away so the check will most likely become local to this fine when I merge that in.

File: includes/exceptions.h
Comment (skottmckay) [reply includes/exceptions.h:43]: It would be good to make all the naming/namespaces consistent in a separate PR. We have `OrtW` (not sure what that means), ORTX_, OCOS_ and various other things in places. We should agree on a macro prefix and namespace and update everything to use those. 

I think that's a separate PR though as it will touch a lot of files.

File: includes/onnxruntime_customop.hpp
Comment (skottmckay) [reply includes/onnxruntime_customop.hpp:70]: Probably. I was contemplating adding new CreateKernel/Compute functions to the ORT CustomOp to allow returning a status when I set it up this way, but as we're just re-throwing or calling abort() it may be fine (if the compiler knows an `abort()` means no return is required).

File: includes/exceptions.h
Comment (wejoncy) [inline includes/exceptions.h:37]: Would fprintf(stderr, ....) be safer?

Comment (skottmckay) [reply includes/exceptions.h:37]: Typically in C++ using IO streams  is safer and the recommended approach as all the arguments are type checked at compile type.

https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#slio3-prefer-iostreams-for-io

There's no performance reason here to use printf.

File: includes/onnxruntime_customop.hpp
Comment (skottmckay) [reply includes/onnxruntime_customop.hpp:70]: Yeah nah. It's not smart enough to treat the 'abort()' as a return path so generates a warning (at least with MSVC)

File: CMakeLists.txt
Comment (skottmckay) [inline CMakeLists.txt:179]: This allows building with _OCOS_PREVENT_EXCEPTION_PROPAGATION_OVERRIDE=ON to simulate an embedded build with exceptions disabled so that the test_exceptions.cc tests can be run. Will look at setting up a CI with this tomorrow unless there's a reason to prefer a full embedded build in the CI.

File: includes/exceptions.h
Comment (skottmckay) [reply includes/exceptions.h:37]: @wenbingl Please let me know how I can re-create the warning from using std::cerr. Can change to printf if we can't fix that warning.

File: CMakeLists.txt
Comment (skottmckay) [reply CMakeLists.txt:167]: Things are inlined now so this shouldn't be an issue

File: includes/exceptions.h
Comment (edgchen1) [reply includes/exceptions.h]: ok, sounds good

File: CMakeLists.txt
Comment (edgchen1) [reply CMakeLists.txt:167]: > Do you have an alternative naming suggestion?

not really, just didn't quite understand the naming conventions yet. now I'm also wondering when the `[_]OCOS` prefix should be used.

> AFAIK cmake doesn't allow returning a value from a function

FWIW, we could have function output variables by providing an output variable name parameter, e.g., SetExceptionsRequired(_OCOS_EXCEPTIONS_REQUIRED).

File: includes/ocos.h
Comment (skottmckay) [reply includes/ocos.h:32]: I agree. I don't think modifying the default value be expected. I would lean towards making it `const T default_value` to make that more explicit as well.

File: CMakeLists.txt
Comment (wenbingl) [inline CMakeLists.txt:167]: Given we only disable exception flags for the non-exception operation, do we still need _OCOS_EXCEPTIONS_REQUIRED since this flag requires the maintenance of the build option list.

File: CMakeLists.txt
Comment (wenbingl) [inline CMakeLists.txt:209]: Do we still need to report the error here? since C++ exceptions can be partially disabled for some operators now.

**Comment by wenbingl** [comment]
@skottmckay , do you still need to refine this PR? 
I will check it later today and approve and merge it if there is no major issue, then cut a release branch.

File: CMakeLists.txt
Comment (skottmckay) [reply CMakeLists.txt:209]: Possibly not, but not sure about one aspects of the setup though.

We're building some operators with and some without exceptions.
We have ORTX_CXX_API_THROW to provide a consistent way to throw, but that is dependent on OCOS_NO_EXCEPTIONS which is not operator specific. 
Operators have no way to report errors in construction or compute with the current ORT custom op API having no ability to return a status, so AFAIK they MUST throw (or log+abort if execeptions disabled) for any error.

What is the expected operator implementation for something that compiles without exceptions? The string operators currently use ORTX_CXX_API_THROW but I think are in the operators we build with exceptions disabled so not sure how that is working.

**Comment by skottmckay** [comment]
> @skottmckay , do you still need to refine this PR? I will check it later today and approve and merge it if there is no major issue, then cut a release branch. have a good weekend!

Maybe some minor things once I test more builds. I want to add something to the CIs but that doesn't necessarily have to block the release if manual testing is OK for checkin. Can we look to cut the release branch first Monday so I can do more testing/validation over the w/e and on my Monday.

File: CMakeLists.txt
Comment (skottmckay) [reply CMakeLists.txt:209]: Is the current setup if exceptions are disabled as follows:
- set OCOS_NO_EXCEPTIONS 
  - which causes all calls to ORTX_CXX_API_THROW to log+abort
- compile the noexcep_operators operators with adjusted flags to disable exceptions
  - this code can use ORTX_CXX_API_THROW which will result in log+abort
- compile other operations with exceptions enabled
  -  3rd party code can call `throw`

For the operators that throw in a build with exception disabled what is the expected behavior? Propagate exception up (no call to `abort`), or something like OCOS_API_IMPL_END should catch the exception and log+abort so all operators have consistent behavior?

File: includes/exceptions.h
Comment (wenbingl) [reply includes/exceptions.h:37]: Some warning like this,
C:\Program Files\Microsoft Visual Studio\2022\Enterprise\VC\Tools\MSVC\14.35.32215\include\ostream(360,1): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc

But, I just figure out defining a _HAS_EXCEPTIONS=0 can remove the warnings from MSVC, but not sure if it is officially supported and how about other compilers,

File: CMakeLists.txt
Comment (wenbingl) [reply CMakeLists.txt:209]: My initial thinking is simple:
If OCOS_ENABLE_CPPEXCEPTION is OFF, then all code in ort-extensions will NOT throw any exception since ORTX-THROW will abort immediately if something unexpected is happening.
But try-catch macro in API function is effective and it actually just only catch the exception from 3rd party libraries.

There is one thing missing here is that BaseKernel should have an exception-free superclass, which will be the base class for non-exception operators. So if only non-exception operators are compiled, there actually no try-catch code in whole compiling source code, I guess the binary size is still minimal in that case.

Comment (skottmckay) [reply CMakeLists.txt:209]: Updated to also log+abort in OCOS_API_IMPL_END if exceptions are disabled.

Not sure it's worth the complexity of a BaseKernel layer without testing to see if the binary size saving justifies it and there a use case that requires it.

**Comment by skottmckay** [comment]
> > @skottmckay , do you still need to refine this PR? I will check it later today and approve and merge it if there is no major issue, then cut a release branch. have a good weekend!
> 
> Maybe some minor things once I test more builds. I want to add something to the CIs but that doesn't necessarily have to block the release if manual testing is OK for checkin. Can we look to cut the release branch first Monday so I can do more testing/validation over the w/e and on my Monday.

PR is ready to be checked in.

**Review by wenbingl — APPROVED** [review]

**Pull request merged**
